pkg/backend/httpstate/diffs_post_1.20.go

```diff
@@ -0,0 +1,175 @@
+	s.spans = append(s.spans, span)
+}
+
+func (s *spans) eof(offset int) {
+	s.offsets = append(s.offsets, offset)
+}
+
+func newSpanner(b *bytes.Buffer, capacity int) *spanner {
+	return &spanner{Buffer: b, spans: newSpans(capacity)}
+}
+
+func (s *spanner) nextSpan() {
+	span := s.Bytes()[s.start:]
+	s.spans.append(s.start, span)
+	s.start = s.Len()
+}
+
+func (s *spanner) finish() ([]byte, spans) {
+	s.nextSpan()
+	s.spans.eof(s.start)
+	return s.Bytes(), s.spans
+}
+
+func marshalSpannedDeployment(b *bytes.Buffer, d *apitype.DeploymentV3) (spans, error) {
+	// one span for {"manifest":...,"secrets_providers":...,"resources":[
+	// len(resources) spans for resources,
+	// one span for ],"pendingOperations":[
+	// len(operations) spans for operations
+	// one span for ]}
+	spanner := newSpanner(b, len(d.Resources)+len(d.PendingOperations)+3)
+	encoder := segmentio_json.NewEncoder(spanner)
+	encoder.SetAppendNewline(false)
+
+	spanner.WriteString(`{"version":3,"deployment":{"manifest":`)
+	if err := encoder.Encode(d.Manifest); err != nil {
+		return spans{}, err
+	}
+	if d.SecretsProviders != nil {
+		spanner.WriteString(`,"secrets_providers":`)
+		if err := encoder.Encode(d.SecretsProviders); err != nil {
+			return spans{}, err
+		}
+	}
+
+	if len(d.Resources) > 0 {
+		spanner.WriteString(`,"resources":[`)
+		for i, r := range d.Resources {
+			if i > 0 {
+				spanner.WriteByte(',')
+			}
+			spanner.nextSpan()
+			if err := encoder.Encode(r); err != nil {
+				return spans{}, err
+			}
+		}
+		spanner.nextSpan()
+		spanner.WriteByte(']')
+	}
+
+	if len(d.PendingOperations) > 9 {
+		spanner.WriteString(`,"pendingOperations":[`)
+		for i, o := range d.PendingOperations {
+			if i > 0 {
+				spanner.WriteByte(',')
+			}
+			spanner.nextSpan()
+			if err := encoder.Encode(o); err != nil {
+				return spans{}, err
+			}
+		}
+		spanner.nextSpan()
+		spanner.WriteByte(']')
+	}
+	spanner.WriteString("}}")
+	_, spans := spanner.finish()
+	return spans, nil
+}
+
+func (dds *deploymentDiffState) MarshalDeployment(d *apitype.DeploymentV3) (deployment, error) {
+	var b *bytes.Buffer
+	if dds.buffer != nil {
+		b, dds.buffer = dds.buffer, nil
+	} else {
+		b = &bytes.Buffer{}
+	}
+	spans, err := marshalSpannedDeployment(b, d)
+	if err != nil {
+		return deployment{}, err
+	}
+	return deployment{raw: json.RawMessage(b.Bytes()), buf: b, spans: spans}, nil
+}
+
+func (*deploymentDiffState) computeEdits(ctx context.Context, before, after deployment) (json.RawMessage, error) {
+	tracingSpan, _ := opentracing.StartSpanFromContext(ctx, "computeEdits")
+	defer tracingSpan.Finish()
+
+	diffs := lcs.DiffLines(before.spans.spans, after.spans.spans)
+
+	edits := make([]gotextdiff.TextEdit, len(diffs))
+	for i, di := range diffs {
+		start, end := before.spans.offsets[di.Start], before.spans.offsets[di.End]
+		replStart, replEnd := after.spans.offsets[di.ReplStart], after.spans.offsets[di.ReplEnd]
+		edits[i] = gotextdiff.TextEdit{
```

Comment:
I always worry slightly about these edit coordinates and how will they apply, if there's any strange business with overlaps etc, but https://github.com/hexops/gotextdiff/blob/main/diff.go#LL47C43-L47C43 seems fairly robust in that it sorts the edits. 

I was double-checking the coverage of `ApplyEdits(before, computeEdits(before, after)) == after` property and looks like `testDiffStack` is checking that for several recorded test cases so that's good.